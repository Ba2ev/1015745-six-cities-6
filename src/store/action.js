export const ActionType = {
  UPDATE_CITY: `city/updateCity`,
  UPDATE_SORT: `sort/updateSort`,
  TOGGLE_SORTLIST: `sort/toggleSortList`,
  UPDATE_HOVERED_OFFER_ID: `offer/updateHoveredOfferId`,
  LOAD_OFFERS: `offers/loadOffers`,
  LOAD_FAVORITES_OFFERS: `offers/loadFavoritesOffers`,
  UPDATE_FAVORITES_OFFERS: `offers/updateFavoritesOffers`,
  LOAD_OFFER_DATA: `offer/loadOfferData`,
  LOAD_OFFER_COMMENTS: `offer/loadOfferComments`,
  LOAD_OFFER_NEARBY: `offer/loadOfferNearby`,
  REQUIRED_AUTHORIZATION: `user/requiredAuthorization`,
  UPDATE_ACCOUNT: `user/updateAccount`,
};

export const ActionCreator = {
  updateCity: (city) => ({
    type: ActionType.UPDATE_CITY,
    payload: city
  }),
  updateSort: (sortType) => ({
    type: ActionType.UPDATE_SORT,
    payload: sortType
  }),
  toggleSortList: () => ({
    type: ActionType.TOGGLE_SORTLIST,
  }),
  updateHoveredOfferId: (id) => ({
    type: ActionType.UPDATE_HOVERED_OFFER_ID,
    payload: id
  }),
  loadOffers: (offers) => ({
    type: ActionType.LOAD_OFFERS,
    payload: offers
  }),
  loadFavoritesOffer: (offers) => ({
    type: ActionType.LOAD_FAVORITES_OFFERS,
    payload: offers
  }),
  loadOfferNearby: (nearblyOffers) => ({
    type: ActionType.LOAD_OFFER_NEARBY,
    payload: nearblyOffers
  }),
  loadOfferData: (offerData) => ({
    type: ActionType.LOAD_OFFER_DATA,
    payload: offerData
  }),
  loadOfferComments: (offerComments) => ({
    type: ActionType.LOAD_OFFER_COMMENTS,
    payload: offerComments
  }),
  updateFavoriteOffers: (offerData) => ({
    type: ActionType.UPDATE_FAVORITES_OFFERS,
    payload: offerData
  }),
  requireAuthorization: (status) => ({
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload: status,
  }),
  updateAccount: (email) => ({
    type: ActionType.UPDATE_ACCOUNT,
    payload: email,
  }),
};

