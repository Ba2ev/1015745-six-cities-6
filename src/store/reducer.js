import {combineReducers} from 'redux';
import {user} from './user/user';
import {property} from './property/property';
import {places} from './places/places';

export const NameSpace = {
  PROPERTY: `PROPERTY`,
  PLACES: `PLACES`,
  USER: `USER`
};

export default combineReducers({
  [NameSpace.PROPERTY]: property,
  [NameSpace.PLACES]: places,
  [NameSpace.USER]: user,
});
