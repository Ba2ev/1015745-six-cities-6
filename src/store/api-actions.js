import {loadOffers, loadOfferData, loadOfferComments, loadOfferNearby, loadFavoritesOffer, updateOfferFavorite, requireAuthorization, updateAccount} from './action';
import {adaptOfferToClient, adaptCommentToClient} from '../adapter/adapter';

export const fetchOffers = () => (dispatch, _getState, api) => (
  api.get(`/hotels`)
    .then(({data}) => (data.map(adaptOfferToClient)))
    .then((data) => dispatch(loadOffers(data)))
);

export const fetchOfferData = (id) => (dispatch, _getState, api) => (
  api.get(`/hotels/${id}`)
    .then(({data}) => (adaptOfferToClient(data)))
    .then((data) => dispatch(loadOfferData(data)))
);

export const fetchOfferComments = (id) => (dispatch, _getState, api) => (
  api.get(`/comments/${id}`)
    .then(({data}) => (data.map(adaptCommentToClient)))
    .then((data) => dispatch(loadOfferComments(data)))
);

export const fetchOfferNearby = (id) => (dispatch, _getState, api) => (
  api.get(`/hotels/${id}/nearby`)
    .then(({data}) => (data.map(adaptOfferToClient)))
    .then((data) => dispatch(loadOfferNearby(data)))
);

export const fetchFavoritesOffer = () => (dispatch, _getState, api) => (
  api.get(`/favorite`)
    .then(({data}) => (data.map(adaptOfferToClient)))
    .then((data) => dispatch(loadFavoritesOffer(data)))
);

export const changeFavorite = ({id, status}) => (dispatch, _getState, api) => (
  api.post(`/favorite/${id}/${status}`)
    .then(({data}) => (adaptOfferToClient(data)))
    .then((data) => dispatch(updateOfferFavorite(data)))
);

export const sendComment = ({id, comment, rating}) => (dispatch, _getState, api) => (
  api.post(`/comments/${id}`, {comment, rating})
    .then(({data}) => (data.map(adaptCommentToClient)))
    .then((data) => dispatch(loadOfferComments(data)))
);

export const checkAuth = () => (dispatch, _getState, api) => (
  api.get(`/login`)
    .then(({data}) => dispatch(updateAccount(data.email)))
    .then(() => dispatch(requireAuthorization(true)))
    .catch(() => {})
);

export const login = ({login: email, password}) => (dispatch, _getState, api) => (
  api.post(`/login`, {email, password})
    .then(() => dispatch(requireAuthorization(true)))
    .then(() => dispatch(updateAccount(email)))
);

export const logout = () => (dispatch, _getState, api) => (
  api.get(`/logout`)
    .then(() => dispatch(requireAuthorization(false)))
    .then(() => dispatch(updateAccount(``)))
);
