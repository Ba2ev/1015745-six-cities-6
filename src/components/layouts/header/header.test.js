import React from 'react';
import {render} from '@testing-library/react';
import {Router} from 'react-router-dom';
import * as redux from 'react-redux';
import configureStore from 'redux-mock-store';
import {createMemoryHistory} from 'history';
import Header from './header';
import {NameSpace} from '../../../store/reducer';

const FAKE_EMAIL = `test@test.ru`;

const mockStore = configureStore({});

jest.spyOn(redux, `useSelector`);
it(`Header should render correctly`, () => {
  const history = createMemoryHistory();

  const initialState = {
    [NameSpace.USER]: {
      isAuthorized: true,
      account: FAKE_EMAIL
    }
  };

  const {getByText} = render(
      <redux.Provider store={mockStore(initialState)}>
        <Router history={history}>
          <Header />
        </Router>
      </redux.Provider>
  );

  const email = getByText(FAKE_EMAIL);

  expect(email).toBeInTheDocument();
});
