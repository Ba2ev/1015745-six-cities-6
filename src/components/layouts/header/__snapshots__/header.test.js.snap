// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render snapshot correctly 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="container"
    >
      <div
        class="header__wrapper"
      >
        <div
          class="header__left"
        >
          <a
            class="header__logo-link"
            href="/"
          >
            <img
              alt="6 cities logo"
              class="header__logo"
              height="81"
              src="img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          class="header__nav"
        >
          <ul
            class="header__nav-list"
          >
            <li
              class="header__nav-item user"
            >
              <a
                class="header__nav-link header__nav-link--profile"
                href="/favorites"
              >
                <div
                  class="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  class="header__user-name user__name"
                >
                  test@test.ru
                </span>
              </a>
              <a>
                Sign out
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
</div>
`;
